plugins/camera/camera_impl.cpp

```diff
@@ -0,0 +1,739 @@
+        if (callback) {
+            callback(Camera::Result::WRONG_ARGUMENT);
+        }
+        return;
+    }
+
+    std::lock_guard<std::mutex> lock(_capture.mutex);
+
+    _parent.send_command_with_ack_async(
+        MAV_CMD_IMAGE_START_CAPTURE,
+        MavlinkCommands::Params {0.0f, // Reserved, set to 0
+                                 interval_s,
+                                 0.0f, // unlimited photos
+                                 float(_capture.sequence++),
+                                 NAN, NAN, NAN},
+        std::bind(&CameraImpl::receive_command_result, _1, callback),
+        MAV_COMP_ID_CAMERA);
+}
+
+void CameraImpl::stop_photo_interval_async(const Camera::result_callback_t &callback)
+{
+    _parent.send_command_with_ack_async(
+        MAV_CMD_IMAGE_STOP_CAPTURE,
+        MavlinkCommands::Params {0.0f, // Reserved, set to 0
+                                 NAN, NAN, NAN, NAN, NAN, NAN},
+        std::bind(&CameraImpl::receive_command_result, _1, callback),
+        MAV_COMP_ID_CAMERA);
+}
+
+void CameraImpl::start_video_async(const Camera::result_callback_t &callback)
+{
+    _parent.send_command_with_ack_async(
+        MAV_CMD_VIDEO_START_CAPTURE,
+        MavlinkCommands::Params {0.0f, // Reserved, set to 0
+                                 NAN, // fps not set yet
+                                 NAN, // resolution not set yet
+                                 NAN, NAN, NAN, NAN},
+        std::bind(&CameraImpl::receive_command_result, _1, callback),
+        MAV_COMP_ID_CAMERA);
+}
+
+void CameraImpl::stop_video_async(const Camera::result_callback_t &callback)
+{
+    _parent.send_command_with_ack_async(
+        MAV_CMD_VIDEO_STOP_CAPTURE,
+        MavlinkCommands::Params {0.0f, // Reserved, set to 0
+                                 NAN, NAN, NAN, NAN, NAN, NAN},
+        std::bind(&CameraImpl::receive_command_result, _1, callback),
+        MAV_COMP_ID_CAMERA);
+}
+
+Camera::Result CameraImpl::camera_result_from_command_result(MavlinkCommands::Result
+                                                             command_result)
+{
+    switch (command_result) {
+        case MavlinkCommands::Result::SUCCESS:
+            return Camera::Result::SUCCESS;
+        case MavlinkCommands::Result::NO_DEVICE:
+        // FALLTHROUGH
+        case MavlinkCommands::Result::CONNECTION_ERROR:
+        // FALLTHROUGH
+        case MavlinkCommands::Result::BUSY:
+            return Camera::Result::ERROR;
+        case MavlinkCommands::Result::COMMAND_DENIED:
+            return Camera::Result::DENIED;
+        case MavlinkCommands::Result::TIMEOUT:
+            return Camera::Result::TIMEOUT;
+        default:
+            return Camera::Result::UNKNOWN;
+    }
+}
+
+void CameraImpl::set_mode_async(Camera::Mode mode, const Camera::mode_callback_t &callback)
+{
+    float mavlink_mode;
+    switch (mode) {
+        case Camera::Mode::PHOTO:
+            mavlink_mode = CAMERA_MODE_IMAGE;
```

Comment:
I'm just curious: Didn't you get auto type conversion error from `uint8_t` to `float` ? (In MSVC at least :) )

Reply:
Hehe, don't know.